[FLINK-25920] Ignore duplicate EOI in SinkWriter

AHeise opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

[FLINK-25920]

In case of a failure after final checkpoint, EOI is called twice. SinkWriter should ignore the second call to avoid emitting duplicate committables. This commit uses a union state to remember that EOI happened and suppress additional handling. 

## Brief change log

* Improve sink test assertions
* Straighten EOI handling in CommittableCollector
* Ignore duplicate EOI in SinkWriter <- main fix
* Improve logging in committable handling of the sink

## Verifying this change

Please make sure both new and modified tests in this PR follow [the conventions for tests defined in our code quality guide](https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#7-testing).

- Added integration tests that covers the fix

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -160,7 +169,19 @@ public void initializeState(StateInitializationContext context) throws Exception
+
+        // use union state to ensure that rescaling works correctly
+        this.endOfInputState =
+                context.getOperatorStateStore().getUnionListState(END_OF_INPUT_STATE_DESC);
```
I need help understanding using the state.

I would expect that after receiving EOI we can also not persist anything anymore.

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -249,13 +276,18 @@ private void emit(
 
+    private void emit(StreamRecord<CommittableMessage<CommT>> message) {
+        output.collect(message);
+        LOG.debug("Sending message to committer: {}", message);
```
Nit: Either move the statement before sending or please do a slight rename.
```suggestion
        LOG.debug("Sent message to committer: {}", message);
```

fapaul reviewed (commented):
Thanks for working on this issue. 

The refactoring of the checkpoint id is very nice 👍 

Regarding the actual fix I am not fully convinced yet. Even if persisting the information on EOI works it feels slightly off.

WDYT about not sending the commitable summary on EOI if the summary is empty? It seems the much easier approach.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -195,12 +216,18 @@ public void processWatermark(Watermark mark) throws Exception {
-        sinkWriter.flush(true);
-        emitCommittables(Long.MAX_VALUE);
+        if (!this.endOfInput) {
+            this.endOfInput = true;
+            // endOfInputState is union state, so it's enough if one task adds something to it
```
What happens to the sinkWriter when two subtasks with different endOfInputState got merged on down-scaling?  Shouldn't sink writer be be flushed? (I see it also uses operator list state ("writer_raw_states")) IIUC, it wouldn't

AHeise replied:
You are correct that this was not correct, especially also on partially completed tasks.

I changed it to regular state and let the committer handle duplicates now. The main idea is to avoid extra transactions as much as possible in the writer and if it's not possible, let the committer clean them up anyhow.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -195,12 +216,18 @@ public void processWatermark(Watermark mark) throws Exception {
+        if (!this.endOfInput) {
+            this.endOfInput = true;
+            // endOfInputState is union state, so it's enough if one task adds something to it
+            if (getRuntimeContext().getTaskInfo().getIndexOfThisSubtask() == 0) {
+                endOfInputState.add(true);
```
What happens if we up-scale after this? 
should we drop all other states?
Otherwise, should we check in initState() all items?

AHeise replied:
I changed the logic entirely. We now drop and readd the state in `snapshotState`. This means if a downscaling happens and EOI_state = [true, false], on init endOfInput = false, and on next checkpoint EOI_state=[false]. I discussed the different cases in code.

rkhachatryan reviewed (commented):
Thanks for the fix @AHeise 

I've left some comments, PTAL.

Besides of that, I think `CompactorOperator.java` needs to be adjusted:
- https://github.com/apache/flink/blob/277706d8cb2d446c2d74cb5170158ecc97859acf/flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperator.java#L149 - this code path still emits null
- https://github.com/apache/flink/blob/277706d8cb2d446c2d74cb5170158ecc97859acf/flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperator.java#L252 possible NPE (or remove `@Nullable`)
- https://github.com/apache/flink/blob/277706d8cb2d446c2d74cb5170158ecc97859acf/flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperator.java#L261 possible NPE (or remove `@Nullable`)

Also, could you clarify what is the exact path when a duplicate summary for the same checkpoint was emitted? IIUC, the 2nd code path is the "normal" `endInput` by `SinkWriterOperator` for then bounded input. But what is the 1st path?

AHeise commented:
> Besides of that, I think `CompactorOperator.java` needs to be adjusted:
Thanks for the pointers. I'll address it later today.

> Also, could you clarify what is the exact path when a duplicate summary for the same checkpoint was emitted? IIUC, the 2nd code path is the "normal" `endInput` by `SinkWriterOperator` for then bounded input. But what is the 1st path?

Let's first recap why we need to emit in EOI at all: For streaming jobs, emitting committables on barrier is the correct behavior. For batch, emitting on EOI.

However, there is an additional case around streaming with bounded input and final checkpoints. The operator first receives an EOI and then the final barrier. After EOI, the operator is not allowed to emit another record. So just for this case, streaming also needs to emit on EOI and suppress on barrier.

Now the question is what happens on failure after final checkpoint. Logically speaking, the EOI before the checkpoint should influence the operator state in the checkpoint in such a way that after recover we are still not allowed to emit records. For sinks that means that all committables have already been transferred to the committer operator for the snapshot. For technical reasons, we still receive a second EOI after recovery. Logically, the first EOI should have sufficed. I'm assuming Flink does the second EOI mostly for channel management but maybe the proper fix is actually not calling EOI twice. @pnowojski could you PTAL?

Roman, you raise good questions concerning rescaling and more. The current implementation assumes that EOI will be received by all subtasks at the same logical time (e.g. on final checkpoint). Can we have instances where some subtasks shutdown earlier?

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:
```diff
@@ -208,8 +209,8 @@ public void processElement(StreamRecord<CommittableMessage<CommT>> element) thro
-        OptionalLong checkpointId = element.getValue().getCheckpointId();
-        if (checkpointId.isPresent() && checkpointId.getAsLong() <= lastCompletedCheckpointId) {
+        long checkpointId = element.getValue().getCheckpointIdOrEOI();
+        if (checkpointId <= lastCompletedCheckpointId) {
```
This is kinda of a change in the semantics on first glance. However, we should not receive any elements after EOI, so this code path is actually never triggered and now it's simpler.

fapaul replied:
I do not fully understand the comment.

- Why is it a change in semantics? The condition looks the same the only case that is removed is a committable without checkpoint. Which scenario was this before the change?
- I thought that for the final checkpoint, we would first receive EOI and then do a final checkpoint. This would mean the committer receives data after EOI.

AHeise replied:
Yes, we mean the same thing. For EOI committables, this check will always yield false. Even without the special case on EOI.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:
```diff
@@ -156,23 +156,16 @@ public void endInput() throws Exception {
-                boolean fullyReceived =
-                        !endInput && manager.getCheckpointId() == lastCompletedCheckpointId;
-                commitAndEmit(manager, fullyReceived);
+                    committableCollector.getCheckpointCommittablesUpTo(completedCheckpointId)) {
```
@fapaul please double-check why we did it in this complicated manner originally.

fapaul replied:
On first glance this change doesn't look correct. 

By removing `fullyReceived`, can we now commit committables that are from the "current" checkpoint but on receival from a delayed notifyCheckpointComplete from a previous checkpoint.

AHeise replied:
Just in case you haven't seen it: `fullyReceived` is now implicitly always true.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:
```diff
@@ -72,18 +73,24 @@ Collection<SubtaskCommittableManager<CommT>> getSubtaskCommittableManagers() {
     }
 
-    void upsertSummary(CommittableSummary<CommT> summary) {
+    void addSummary(CommittableSummary<CommT> summary) {
```
Changed the name as upsert imples that we always merge but we really just want to do this for EOI now.

AHeise commented:
I addressed your comments and took a different approach to state management (it looks similar but is conceptually rather different).

I restructured the PR and added 2 more commits to it. Unfortunately, the fixups looked then rather messy so I decided to force push everything again. So I'm sorry, but you more or less have to review again ;).

Btw I figured that the EOI changes were a breaking change of the experimental CommittableMessage and made them non-breaking instead. So the commit is essentially the same but looks a bit different.

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:
```diff
@@ -147,15 +146,16 @@ Collection<CommittableWithLineage<CommT>> drainFinished() {
 
     CheckpointCommittableManagerImpl<CommT> merge(CheckpointCommittableManagerImpl<CommT> other) {
-        checkArgument(Objects.equals(other.checkpointId, checkpointId));
+        checkArgument(other.checkpointId == checkpointId);
```
Nit: This change should probably go into the commit, changing the type of the checkpoint fields

AHeise replied:
Good catch. I'll try to move it.

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -160,13 +171,38 @@ public void initializeState(StateInitializationContext context) throws Exception
-        sinkWriter = writerStateHandler.createWriter(initContext, context);
+
+        endOfInputState = context.getOperatorStateStore().getListState(END_OF_INPUT_STATE_DESC);
+        // We can have the following cases:
+        // 1. state is empty:
+        //   - First time initialization
+        //   - Restoring from a previous version of Flink
+        //   - Upscaled from a final or regular checkpoint
+        // In all cases, we properly initialize the data and handle EOI, potentially resulting in
+        // duplicate summaries that the CommitterOperator needs to handle.
+        // 2. state is not empty:
+        //   - This implies Flink restores from a version that handles EOI
+        //   - If there is one entry, no rescaling happened (for this subtask), so if it's true, we
+        //     recover from a final checkpoint (for this subtask) and can ignore another EOI else we
+        //     have a regular checkpoint.
+        //   - If there are multiple entries, Flink downscaled, and we need to check if all are true
+        //     and do the same as above. As soon as one entry is EOI, we regularly start the writer
+        //     and potentially emit duplicate summaries if we indeed recovered from a final
+        //     checkpoint.
```
Very nice 👍

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -160,13 +171,38 @@ public void initializeState(StateInitializationContext context) throws Exception
+        this.endOfInput = !previousState.isEmpty() && !previousState.contains(false);
+        sinkWriter =
+                this.endOfInput
+                        ? new ClosedWriter<>()
```
As discussed offline, this probably leaves an unclean state from the SinkWriter when it previously crashed.

AHeise replied:
Fixed in a fixup commit. PTAL.

fapaul commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -178,7 +214,7 @@ public void processElement(StreamRecord<InputT> element) throws Exception {
     public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
         super.prepareSnapshotPreBarrier(checkpointId);
-        if (!endOfInput) {
+        if (!this.endOfInput) {
```
Nit: Why `this` here we do not use inside the other methods?

AHeise replied:
Yes, I'll remove.

fapaul reviewed (commented):
Overall looks good but I left a few comments which we should answer.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:
```diff
@@ -195,12 +231,15 @@ public void processWatermark(Watermark mark) throws Exception {
-        endOfInput = true;
-        sinkWriter.flush(true);
-        emitCommittables(Long.MAX_VALUE);
+        if (!endOfInput) {
+            endOfInput = true;
```
Would it make sense to disallow `SinkWriterOperator` to `processElement` anymore after this?

AHeise replied:
Interesting idea but wouldn't we validate the framework then?

Or would you also add the assertion after recovery? That's kinda done with the `ClosedSinkWriter`.

AHeise replied:
I added the assertion. It can't hurt because we may have upstream operators that misbehave (custom operators).

rkhachatryan reviewed (approved):
LGTM, the fix should work and I don't see any issues with it.

AHeise commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperator.java:
```diff
@@ -138,15 +136,15 @@ public void processElement(StreamRecord<CompactorRequest> element) throws Except
         assert checkpointRequests.isEmpty();
 
         getAllTasksFuture().join();
-        emitCompacted(null);
```
Afaik yes. We pass the param directly to both `CommittableMessage`s. The serializer replaces `null` with `EOI`. So it effectively results in the same bytes.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/CommittableMessageSerializer.java:
```diff
@@ -91,13 +90,13 @@ public CommittableMessage<CommT> deserialize(int version, byte[] serialized)
                         SimpleVersionedSerialization.readVersionAndDeSerialize(
                                 committableSerializer, in),
-                        readCheckpointId(in),
+                        in.readLong(),
```
Thanks for challenging that. It's one of the parts where most uncertainty still resides. But let's look again on what is happening and has happened:
* On write: we replaced null with EOI, now we get EOI and write it; so it should be the same byte sequence.
* On read: we have always used `readLong` which is not null-aware. We have replaced EOI with null.

So for serialization nothing has changed. However, around serialization we replace null with EOI in all instances of the Message.

That should be safe to change without migration, no?

For compatibility, I left getCheckpointId the same, so it will return an OptionalLong.empty() on EOI where it previously returned it on null. I have not found any usage of the method outside of Flink anyhow.

AHeise commented:
I added the fixup commits inline. PTAL.

fapaul reviewed (approved):
Thanks for walking me through the changes offline and discussing the last open points 👍

fapaul commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/stream/AbstractStreamingWriter.java:
```diff
@@ -149,15 +190,20 @@ public void processElement(StreamRecord<IN> element) throws Exception {
-        commitUpToCheckpoint(Long.MAX_VALUE);
+        if (!this.endOfInput) {
+            this.endOfInput = true;
+            buckets.onProcessingTime(Long.MAX_VALUE);
+            helper.snapshotState(Long.MAX_VALUE);
+            output.emitWatermark(new Watermark(Long.MAX_VALUE));
+            commitUpToCheckpoint(Long.MAX_VALUE);
```
Can we use the `EOI` variable instead of `Long.MAX_VALUE`.

AHeise replied:
I thought about that but decided against it. This is not related to CommittableSummary directly, so it would feel weird to import it just for the EOI. And someone wise said that this class is deprecated anyways, so I didn't want to change too much.

fapaul reviewed (approved):
Reviewed commit for `Fix AbstractStreamingWriter sending after EOI`

AHeise commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/stream/AbstractStreamingWriter.java:
```diff
@@ -123,6 +142,27 @@ public void bucketInactive(Bucket<IN, String> bucket) {
+        //     the writer and potentially emit duplicate summaries if we indeed recovered from a
+        //     final checkpoint.
+        endOfInputState = context.getOperatorStateStore().getListState(END_OF_INPUT_STATE_DESC);
+        List<Boolean> previousState = Lists.newArrayList(endOfInputState.get());
```
Sure if you drop Java8 support first ;).

AHeise commented:
@flinkbot run azure
